README.rst
========
Overview
========

.. start-badges

.. list-table::
    :stub-columns: 1

    * - tests
      - | |travis|
    * - package
      - | |version| |wheel| |supported-versions| |supported-implementations|

.. |travis| image:: https://api.travis-ci.org/jwbargsten/python-timethese.svg?branch=master
    :alt: Travis-CI Build Status
    :target: https://travis-ci.org/jwbargsten/python-timethese

.. |version| image:: https://img.shields.io/pypi/v/timethese.svg
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/timethese

.. |wheel| image:: https://img.shields.io/pypi/wheel/timethese.svg
    :alt: PyPI Wheel
    :target: https://pypi.org/project/timethese

.. |supported-versions| image:: https://img.shields.io/pypi/pyversions/timethese.svg
    :alt: Supported versions
    :target: https://pypi.org/project/timethese

.. |supported-implementations| image:: https://img.shields.io/pypi/implementation/timethese.svg
    :alt: Supported implementations
    :target: https://pypi.org/project/timethese


.. end-badges

Tired of writing the same code again and again when comparing the runtime of
more than one function? `timethese` helps with this type of micro-benchmarking.
It basically runs timeit (or actually repeat) on multiple functions and spits
out a report.

In one sentence: timethese is timeit for multiple functions with better reporting

* Free software: MIT License

Installation
============

::

    pip install timethese

You can also install the in-development version with::

    pip install https://github.com/jwbargsten/python-timethese/archive/master.zip


Usage
=====

Microbenchmark
--------------

`timethese` has a 3 step approach:

1. define the functions you want to compare
2. feed them to cmpthese as list or dict (see below)
3. format the results, aka pretty print

Let's have a look::

      from timethese import cmpthese, pprint_cmp, timethese

      xs = range(10)


      # 1. DEFINE FUNCTIONS

      def map_hex():
          list(map(hex, xs))


      def list_compr_hex():
          list([hex(x) for x in xs])


      def map_lambda():
          list(map(lambda x: x + 2, xs))


      def map_lambda_fn():
          fn = lambda x: x + 2
          list(map(fn, xs))


      def list_compr_nofn():
          list([x + 2 for x in xs])


      # 2. FEED THE FUNCTIONS TO CMPTHESE

      # AS DICT:

      cmp_res_dict = cmpthese(
          10000,
          {
              "map_hex": map_hex,
              "list_compr_hex": list_compr_hex,
              "map_lambda": map_lambda,
              "map_lambda_fn": map_lambda_fn,
              "list_compr_nofn": list_compr_nofn,
          },
          repeat=3,
      )


      # OR AS LIST:

      cmp_res_list = cmpthese(
          10000, [map_hex, list_compr_hex, map_lambda, map_lambda_fn, list_compr_nofn,], repeat=3,
      )

      # 3. PRETTY PRINT THE RESULTS

      print(pprint_cmp(cmp_res_dict))
      print(pprint_cmp(cmp_res_list))

What do you get if you run this?

Depending on the runtime of the supplied functions, either rate (unit: 1/s) or
the seconds per iteration (s/iter) are shown.

For dict something like::

                            Rate  list_compr_nofn  map_hex  map_lambda  map_lambda_fn  list_compr_hex
      list_compr_nofn  1385057/s                .      43%         47%            48%             88%
              map_hex   969501/s             -30%        .          3%             4%             31%
           map_lambda   940257/s             -32%      -3%           .             1%             27%
        map_lambda_fn   935508/s             -32%      -4%         -1%              .             27%
       list_compr_hex   738367/s             -47%     -24%        -21%           -21%               .

For list something like::

                              Rate  4.list_compr_nofn  0.map_hex  2.map_lambda  3.map_lambda_fn  1.list_compr_hex
      4.list_compr_nofn  1360009/s                  .        31%           42%              46%               78%
              0.map_hex  1037581/s               -24%          .            9%              11%               36%
           2.map_lambda   955513/s               -30%        -8%             .               2%               25%
        3.map_lambda_fn   933666/s               -31%       -10%           -2%                .               22%
       1.list_compr_hex   763397/s               -44%       -26%          -20%             -18%                 .


(the function names are taken from `fn.__name__` and prefixed with the list index.)

Timing
------

`timethese` also has the function `timethese`, which is used by `cmpthese`
internally. To get the timings directly, you can run::

      from timethese import timethese

      xs = range(10)


      def map_hex():
          list(map(hex, xs))


      def list_compr_hex():
          list([hex(x) for x in xs])


      def map_lambda():
          list(map(lambda x: x + 2, xs))


      def map_lambda_fn():
          fn = lambda x: x + 2
          list(map(fn, xs))


      def list_compr_nofn():
          list([x + 2 for x in xs])


      timings_dict = timethese(
          10000,
          {
              "map_hex": map_hex,
              "list_compr_hex": list_compr_hex,
              "map_lambda": map_lambda,
              "map_lambda_fn": map_lambda_fn,
              "list_compr_nofn": list_compr_nofn,
          },
          repeat=3,
      )

      timings_list = timethese(
          10000,
          [ map_hex, list_compr_hex, map_lambda, map_lambda_fn, list_compr_nofn ],
          repeat=3,
      )

      # if you want, you can create a pandas df from it

      import pandas as pd

      timings_df = pd.DataFrame(timings_dict.values())
      print(timings_df)

      # BEWARE: if you pass a list to timings, you have to skip the .values() call

      timings_df = pd.DataFrame(timings_list)
      print(timings_df)


Timing functions with decorators
--------------------------------
`timethese` also provides decorators to time single functions::

     import time
     import timethese

     @timethese.print_time
     def calculate_something():
         time.sleep(1)

     calculate_something()

Four decorators are provided, 2 for normal stuff

* ``timethese.print_time``
* ``timethese.log_time(logger, level=logging.INFO)``

and 2 for pandas dataframes (they also print the shape of the resulting dataframe).
Useful when using ``df.pipe(...)``

* ``timethese.log_time_df(logger, level=logging.INFO)``
* ``timethese.print_time_df``

E.g. to log execution times of pipe operations on pandas dataframes, you could write::

     import time
     import logging
     import timethese
     import numpy as np
     import pandas as pd

     logging.basicConfig(level=logging.DEBUG)

     logger = logging.getLogger(__name__)


     @timethese.log_time_df(logger, logging.DEBUG)
     def sum_by_group(df):
         time.sleep(1)  # introduce some artificial delay
         return df.groupby("A").sum()


     df = pd.DataFrame({"A": np.arange(100) % 2, "B": np.random.normal(size=100)})

     res = df.pipe(sum_by_group)


See the function documentation in the source code for better examples.

Development
===========

To run the all tests run::

    tox

Note, to combine the coverage data from all the tox environments run:

.. list-table::
    :widths: 10 90
    :stub-columns: 1

    - - Windows
      - ::

            set PYTEST_ADDOPTS=--cov-append
            tox

    - - Other
      - ::

            PYTEST_ADDOPTS=--cov-append tox

See also
========

The idea came from Perl's `Benchmark.pm <https://perldoc.perl.org/Benchmark.html>`_, which I used a lot in the Good Ol' Days.


CHANGELOG.rst

Changelog
=========

0.0.7 (2020-05-31)
------------------

* Improved documentation and fixed typos

0.0.6 (2020-05-31)
------------------

* improved documentation
* fixed setup.py install dependencies to again pass travis tests

0.0.5 (2020-05-31)
------------------

* Added better documentations
* Now using NumPy documentation format for function def doc
* Fixed typos in pyproject.toml

0.0.4 (2020-05-30)
------------------

* Fixed code to be compatible with python 3.5
* Fixed travis stuff
* Added decorators to time specific functions for pandas.DataFrame.pipe arguments


0.0.3 (2020-05-27)
------------------

* First release on PyPI.


tests/test_timethese.py
import time

from pytest import approx

from timethese import cmpthese, pprint_cmp, timethese


def test_timethese():
    def a():
        time.sleep(0.3)

    def b():
        time.sleep(0.6)

    res = timethese(2, {"a": a, "b": b})

    assert approx(res["a"]["best_time"], abs=0.1) == 0.3
    assert approx(res["b"]["best_time"], abs=0.1) == 0.6


def test_cmpthese():
    def a():
        time.sleep(0.1)

    def b():
        time.sleep(0.2)

    res = cmpthese(2, {"a": a, "b": b}, repeat=2)
    assert approx(res["data"][0], abs=0.2) == [0, 1]
    assert approx(res["data"][1], abs=0.2) == [-0.5, 0]


def test_pprint_iter():
    res = {"data": [[0, 0.99], [-0.49, 0]], "names": ["a", "b"], "times": [0.10, 0.20], "rates": [9.78, 4.93]}
    expected = """     Rate     a    b
a  9.78/s     .  99%
b  4.93/s  -49%    ."""

    assert pprint_cmp(res) == expected


def test_pprint_rate():
    res = {
        "data": [[0, 0.7415, 84.8878], [-0.4258, 0, 48.3165], [-0.9883, -0.9797, 0]],
        "names": ["b", "c", "a"],
        "times": [0.0011, 0.0020, 0.1015],
        "rates": [845.3621, 485.4044, 9.8426],
    }
    expected = """     Rate     b     c      a
b   845/s     .   74%  8489%
c   485/s  -43%     .  4832%
a  9.84/s  -99%  -98%      ."""
    assert pprint_cmp(res) == expected


ci/bootstrap.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import, print_function, unicode_literals

import os
import subprocess
import sys
from os.path import abspath, dirname, exists, join

base_path = dirname(dirname(abspath(__file__)))


def check_call(args):
    print("+", *args)
    subprocess.check_call(args)


def exec_in_env():
    env_path = join(base_path, ".tox", "bootstrap")
    if sys.platform == "win32":
        bin_path = join(env_path, "Scripts")
    else:
        bin_path = join(env_path, "bin")
    if not exists(env_path):
        import subprocess

        print("Making bootstrap env in: {0} ...".format(env_path))
        try:
            check_call([sys.executable, "-m", "venv", env_path])
        except subprocess.CalledProcessError:
            try:
                check_call([sys.executable, "-m", "virtualenv", env_path])
            except subprocess.CalledProcessError:
                check_call(["virtualenv", env_path])
        print("Installing `jinja2` into bootstrap environment...")
        check_call([join(bin_path, "pip"), "install", "jinja2", "tox"])
    python_executable = join(bin_path, "python")
    if not os.path.exists(python_executable):
        python_executable += ".exe"

    print("Re-executing with: {0}".format(python_executable))
    print("+ exec", python_executable, __file__, "--no-env")
    os.execv(python_executable, [python_executable, __file__, "--no-env"])


def main():
    import jinja2

    print("Project path: {0}".format(base_path))

    jinja = jinja2.Environment(
        loader=jinja2.FileSystemLoader(join(base_path, "ci", "templates")),
        trim_blocks=True,
        lstrip_blocks=True,
        keep_trailing_newline=True,
    )

    tox_environments = [
        line.strip()
        # 'tox' need not be installed globally, but must be importable
        # by the Python that is running this script.
        # This uses sys.executable the same way that the call in
        # cookiecutter-pylibrary/hooks/post_gen_project.py
        # invokes this bootstrap.py itself.
        for line in subprocess.check_output(
            [sys.executable, "-m", "tox", "--listenvs"], universal_newlines=True
        ).splitlines()
    ]
    tox_environments = [line for line in tox_environments if line.startswith("py")]

    for name in os.listdir(join("ci", "templates")):
        with open(join(base_path, name), "w") as fh:
            fh.write(jinja.get_template(name).render(tox_environments=tox_environments))
        print("Wrote {}".format(name))
    print("DONE.")


if __name__ == "__main__":
    args = sys.argv[1:]
    if args == ["--no-env"]:
        main()
    elif not args:
        exec_in_env()
    else:
        print("Unexpected arguments {0}".format(args), file=sys.stderr)
        sys.exit(1)


src/timethese.py
import functools
import logging
import statistics as stat
import time
import timeit

from more_itertools import quantify

__version__ = "0.0.3"


def print_time_df(fn):
    """Decorator to print run time when piping pandas DataFrames

    In addition to the run time, it also prints the shape of the resulting dataframe.

    Examples
    --------
    A common usage looks like this::

        import time
        import timethese
        import numpy as np
        import pandas as pd


        @timethese.print_time_df
        def sum_by_group(df):
            time.sleep(1)  # introduce some artificial delay
            return df.groupby("A").sum()


        df = pd.DataFrame({"A": np.arange(100) % 2, "B": np.random.normal(size=100)})

        res = df.pipe(sum_by_group)

    """
    @functools.wraps(fn)
    def fn_exec_time(df, *args, **kwargs):
        start = time.perf_counter()
        value = fn(df, *args, **kwargs)
        end = time.perf_counter()
        elapsed = end - start
        print("{0} took={1:.4f}s shape={2}".format(fn.__name__, elapsed, value.shape))
        return value

    return fn_exec_time


def print_time(fn):
    """Decorator to print the run time of a function

    Examples
    --------
    A common usage looks like this::

        import time
        import timethese

        @timethese.print_time
        def calculate_something():
            time.sleep(1)

        calculate_something()

    """
    @functools.wraps(fn)
    def fn_exec_time(*args, **kwargs):
        start = time.perf_counter()
        value = fn(*args, **kwargs)
        end = time.perf_counter()
        elapsed = end - start
        print("{0} took={1:.4f}s".format(fn.__name__, elapsed))
        return value

    return fn_exec_time


def log_time(logger, level=logging.INFO):
    """Decorator to log the run time of a function

    Parameters
    ----------
    logger : logging.Logger
        a logger to write the messages to
    level : int
        a logging level to use, e.g. ``logging.DEBUG``. The default level is ``logging.INFO``

    Examples
    --------
    A common usage looks like this::

        import time
        import logging
        import timethese

        logging.basicConfig(level=logging.DEBUG)

        logger = logging.getLogger(__name__)

        @timethese.log_time(logger, level=logging.INFO)
        def calculate_something():
            time.sleep(1)

        calculate_something()
    """
    def wrap(fn):
        @functools.wraps(fn)
        def fn_exec_time(*args, **kwargs):
            start = time.perf_counter()
            value = fn(*args, **kwargs)
            end = time.perf_counter()
            elapsed = end - start
            logger.log(level, "{0}: took={1:.4f}s".format(fn.__name__, elapsed))
            return value

        return fn_exec_time

    return wrap


def log_time_df(logger, level=logging.INFO):
    """Decorator to log run time when piping pandas DataFrames

    In addition to the run time, it also logs the shape of the resulting dataframe.

    Parameters
    ----------
    logger : logging.Logger
        a logger to write the messages to
    level : int
        a logging level to use, e.g. ``logging.INFO``

    Examples
    --------

        import time
        import logging
        import timethese
        import numpy as np
        import pandas as pd

        logging.basicConfig(level=logging.DEBUG)

        logger = logging.getLogger(__name__)


        @timethese.log_time_df(logger, logging.DEBUG)
        def sum_by_group(df):
            time.sleep(1)  # introduce some artificial delay
            return df.groupby("A").sum()


        df = pd.DataFrame({"A": np.arange(100) % 2, "B": np.random.normal(size=100)})

        res = df.pipe(sum_by_group)
    """
    def wrap(fn):
        @functools.wraps(fn)
        def fn_exec_time(df, *args, **kwargs):
            start = time.perf_counter()
            value = fn(df, *args, **kwargs)
            end = time.perf_counter()
            elapsed = end - start
            logger.log(level, "{0} took={1:.4f}s shape={2}".format(fn.__name__, elapsed, value.shape))
            return value

        return fn_exec_time

    return wrap


def _timeit(n, args, repeat=1):
    if isinstance(args, dict):
        kwargs = args
    else:
        kwargs = {"stmt": args}
    run_times = timeit.repeat(**kwargs, repeat=repeat, number=n)
    times = [dt / n for dt in run_times]
    rates = [1 / (t + 0.000000000000001) for t in times]

    return {
        "n": n,
        "nruns": repeat,
        "run_times": run_times,
        "times": times,
        "best_time": min(times),
        "mean": stat.mean(times),
        "median": stat.median(times),
        "stdev": stat.stdev(times) if len(times) >= 2 else 0,
        "rates": rates,
        "best_rate": max(rates),
    }


def _format_perf(perf, is_rate):
    # We assume that we'll never get a 0 rate.
    if perf >= 100:
        fmt = "{:0.0f}".format(perf)
    elif perf >= 10:
        fmt = "{:0.1f}".format(perf)
    elif perf >= 1:
        fmt = "{:0.2f}".format(perf)
    elif perf >= 0.1:
        fmt = "{:0.3f}".format(perf)
    else:
        fmt = "{:0.2e}".format(perf)

    if is_rate:
        return fmt + "/s"

    return fmt


def _fmt_stmt(i, stmt):
    if callable(stmt):
        return "{0}.{1}".format(i, stmt.__name__)
    else:
        return "{0}.{1}".format(i, str(stmt)[:11])


def timethese(n=1, funcs=None, repeat=3):
    """Time multiple functions

    Parameters
    ----------
    n
        the number of times the function is executed

    funcs
        the statements/functions or arguments supplied to timeit (timeit is
        used internally, the elements of funcs are mapped to the `stmt` parameter of
        timeit). `funcs` can be a list ``[stmt1, stmt2, ...]``, a dict
        ``{ "stmt1": stmt1, "stmt2": stmt2, ...}``. If you need advanced functionality,
        you can also supply a dict which is used as kwargs for timeit instead of
        the statement itself: ``[{stmt:'...', setup:'...', timer: ... }, ...]``.

    repeat
        times to repeat the complete performance measurement. This can be
        useful to mitigate "noise" created by other programs. The minimum time
        is taken from the repeats.

    Returns
    -------
    dict
        the run time results of each function/statement. Keys of the dict are:

        * *n* - number of loops per run/repeat
        * *nruns* - number of runs/repeats
        * *run_times* - time of each run (called `repeat` in timeit)
        * *times* - ``[run_time/n for run_time in run_times]``
        * *best_time* - ``min(times)``
        * *mean* - ``stat.mean(times)``
        * *median* - ``stat.median(times)``
        * *stdev* - ``stat.stdev(times) if len(times) >= 2 else 0``
        * *rates* - ``[1/time for time in times]``
        * *best_rate* - ``max(rates)``
    """
    if funcs is None:
        return None
    if isinstance(funcs, dict):
        return {name: {"name": name, **_timeit(n, args, repeat=repeat)} for name, args in funcs.items()}
    elif isinstance(funcs, list):
        return [
            {"name": _fmt_stmt(i, args), **_timeit(n, args, repeat=repeat)} for i, args in enumerate(funcs)
        ]
    else:
        raise TypeError("Unknown type of funcs parameter.")


def cmpthese(n=1, funcs=None, repeat=3):
    """Compare run time of multiple functions.

    Parameters
    ----------
    n
        the number of times the function is executed

    funcs
        the statements/functions or arguments supplied to timeit (timeit is
        used internally, the elements of funcs are mapped to the `stmt` parameter of
        timeit). `funcs` can be a list ``[stmt1, stmt2, ...]``, a dict
        ``{ "stmt1": stmt1, "stmt2": stmt2, ...}``. If you need advanced functionality,
        you can also supply a dict which is used as kwargs for timeit instead of
        the statement itself: ``[{stmt:'...', setup:'...', timer: ... }, ...]``.

    repeat
        times to repeat the complete performance measurement. This can be
        useful to mitigate "noise" created by other programs. The minimum time
        is taken from the repeats.

    Returns
    -------
    dict
        the comparison results as dict. The dict contains following keys:

        * *data* - a matrix (list of lists) with the comparison results, a comparison
          results of 1 indicates 100% faster, -.5 means 50% slower
        * *names* - list of column/row names (column names == row names)
        * *times* - the best time of each func
        * *rates* - the best rate of each func, rate = 1/time
    """
    results = timethese(n=n, funcs=funcs, repeat=repeat)
    if isinstance(funcs, dict):
        results = results.values()

    results = sorted(results, key=lambda r: r["best_time"])
    names = [r["name"] for r in results]
    times = [r["best_time"] for r in results]
    rates = [r["best_rate"] for r in results]

    rows = []
    for row_idx, row_res in enumerate(results):
        row = []

        # Column 1 = performance
        row_rate = row_res["best_rate"]

        for col_idx, col_res in enumerate(results):
            if row_idx == col_idx:
                out = 0
            else:
                col_rate = col_res["best_rate"]
                out = row_rate / col_rate - 1

            row.append(out)
        rows.append(row)
    return {"data": rows, "names": names, "times": times, "rates": rates}


def pprint_cmp(result, as_rate=None):
    """Pretty print a result from timethese

    Parameters
    ----------
    result
        a result from timethese
    as_rate
        If None, the display format is determined automatically.
        If True, a rate is shown as function performance (unit 1/s),
        if False, the time is shown (unit s)

    Returns
    -------
    dict
        a string table with the results
    """
    if as_rate is None:
        # determine rate display automatically
        big_rates_count = quantify(result["times"], lambda t: t < 1)
        as_rate = result and big_rates_count / len(result["times"]) > 0.5

    perf = result["rates"] if as_rate else result["times"]
    perf = [_format_perf(p, as_rate) for p in perf]

    names = result["names"]
    top_row = ["", "Rate" if as_rate else "s/iter", *names]
    rows = [top_row] + [list(x) for x in zip(names, perf)]

    col_widths = [len(x) for x in top_row]
    col_widths[0] = max([len(n) for n in names])
    col_widths[1] = max([len(rows[i][1]) for i in range(len(rows))])

    for ridx, r in enumerate(result["data"]):
        for cidx, c in enumerate(r):
            if ridx == cidx:
                val = "."
            else:
                val = "{:.0f}%".format(c * 100)

            rows[ridx + 1].append(val)

            if len(val) > col_widths[cidx + 2]:
                col_widths[cidx + 2] = len(val)

    def format_row(row, widths):
        return "  ".join(["{:>{fill}s}".format(x[0], fill=x[1]) for x in zip(row, widths)])

    return "\n".join([format_row(r, col_widths) for r in rows])
